src/Chapter3.hs

```diff
@@ -480,6 +528,60 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle String 
+    deriving (Show, Eq)
+
+data Wall = Wall Bool 
+    deriving (Show, Eq)
+
+data Community = Church | Library 
+    deriving (Show, Eq)
+
+data Housing = Housing
+    { oneOccupant    :: Int
+    , twoOccupants   :: Int
+    , threeOccupants :: Int
+    , fourOccupants  :: Int
+    } deriving (Show, Eq)
+
+data City = City
+    { castle    :: Maybe Castle
+    , wall      :: Wall
```

Comment:
This is the correct data type, that can represent what we have in requirements 👍🏼 
However, this also allows having some states that should not ideally exist: like having walls without any castle at all. With this representation, this is possible. However, we actually can restrict this on type-level by providing more representative types 🙂 This is actually the greatest power of Haskell and type system 😍 

This pattern is called "Make illegal states unrepresentable", and you can read more about it here:

* https://kowainik.github.io/posts/haskell-mini-patterns#make-illegal-states-unrepresentable